### PR TITLE
PR: list recordables

### PR DESCRIPTION
**Pull request by sanjayankur31**
This should pass correctly :crossed_fingers:

### CODE REVIEW

**Comment by sanjayankur31** [comment]
@pgleeson : ready now---looks like some conflicts had popped up in the other PR

File: tests/test_pynml.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+            expnames.append(exp.name)
+        # https://docs.neuroml.org/Userdocs/Schemas/Cells.html#schema-izhikevich2007cell
+        self.assertTrue("u" in expnames)
+        self.assertTrue("v" in expnames)
+        self.assertTrue("iMemb" in expnames)
+        self.assertTrue("iSyn" in expnames)
+
+    def test_recording_path_listing(self):
+        """Test listing of recording paths in NeuroML documents."""
+        paths = list_recording_paths("tests/izhikevich_test_file.nml", "iz")
```
Comment (pgleeson) [inline tests/test_pynml.py:70]: Just checking, does it list also the current from the input as recordable, e.g. https://github.com/NeuroML/NeuroML2/blob/master/LEMSexamples/LEMS_NML2_Ex2_Izh.xml#L58

Comment (sanjayankur31) [reply tests/test_pynml.py:70]: I wouldn't think so---`I` is not an exposure in the Izhikevich model:
https://docs.neuroml.org/Userdocs/Schemas/Cells.html#schema-izhikevich2007cell

`I` is an exposure in the pulsegeneratorDL component, so it'll be listed and can be recorded from there:
https://docs.neuroml.org/Userdocs/Schemas/Inputs.html?highlight=pulsegeneratordl#pulsegeneratordl

I don't quite understand why the example is recording the current for the pulsegenerator by referring to the izhikevich cell. Doesn't `izpopBurst[0]/i0/I` suggest that the pulsegenerator is a children element of the cell, which it isn't?

Comment (pgleeson) [reply tests/test_pynml.py:70]: Ah... that's the difference between the LEMS model as specified in the "recipe" to build the network (the LEMS file) and the built network in memory that's run. For the much of the model there is a one to one mapping between entities in the LEMS and what's in the final built model, but in actual fact, what explicitInput is saying is "put an instance of this component on the cell when you build the model", hence  izpopBurst[0]/i0/I is valid when the model is built. 

Similarly, a connection in a projection is saying "put a synapse on this cell and connect it to the event output", so the synapse component on that cell can be recorded from in the final model, see https://github.com/NeuroML/NeuroML2/blob/master/LEMSexamples/LEMS_NML2_Ex6_NMDA.xml#L89, as opposed to recording teh synapse on something like projection0/connection2/synapse1/i

In theory the list_recording_paths could be clever enough to know what elements will be in the final built model ("ah, there's an inputList, so there will be a pulse generator on cells xyz..."), but that's a bit more involved than just parsing the structure of the lems and asking what exposures they have...

Comment (sanjayankur31) [reply tests/test_pynml.py:70]: OK, I get that but then we have a few problems:

- how does one (a user or a dev) know about this? I don't think this is documented anywhere at all. Not even in the lems docs. (I mean I didn't know this until now :laughing: )
- Do all the LEMS implementations do this? (jlems/pylems/eden?): this is simple to verify.
- If they do---how is this information exposed, if it is exposed at all to the user. I.e., what should one look for? These aren't added as new exposures, so what should one look for?

The new pylems methods that get exposures do go over the full constructed model that resides in memory, and so do these helper functions. We're not text parsing the LEMS file and figuring these things out---we're constructing the full model in PyLEMS and then getting exposures from the constructed model. So if this information is exposed, we can add helper methods to make them available to users too.

**Comment by sanjayankur31** [comment]
Updated the function to clearly say that we only list paths for exposures at the moment.

**Comment by sanjayankur31** [comment]
Requires https://github.com/LEMS/pylems/pull/60 to be merged, since the function signature changed.


We now get this sort of output:

```
pytest -k test_recording_path_listing -s
============================================================== test session starts ===============================================================
platform linux -- Python 3.9.6, pytest-6.2.4, py-1.10.0, pluggy-0.13.1
rootdir: /home/asinha/Documents/02_Code/00_mine/2020-OSB/NeuroML-software/pyNeuroML
collected 5 items / 3 deselected / 2 selected

tests/test_pynml.py pyNeuroML >>> pyneuroml.pynml - INFO - NeuroML LEMS definition files extracted to: /tmp/tmpyx5boi_w/NeuroML2CoreTypes/
IzhNet/IzhPop0[0]/iMemb
IzhNet/IzhPop0[0]/iSyn
IzhNet/IzhPop0[0]/u
IzhNet/IzhPop0[0]/v
IzhNet/explicitInput/pulseGen_0/i
.pyNeuroML >>> pyneuroml.pynml - INFO - NeuroML LEMS definition files extracted to: /tmp/tmpi1nbq6hx/NeuroML2CoreTypes/
single_hh_cell_network/pop0[0]/caConc
single_hh_cell_network/pop0[0]/caConcExt
single_hh_cell_network/pop0[0]/iCa
single_hh_cell_network/pop0[0]/iChannels
single_hh_cell_network/pop0[0]/iSyn
single_hh_cell_network/pop0[0]/spiking
single_hh_cell_network/pop0[0]/surfaceArea
single_hh_cell_network/pop0[0]/totSpecCap
single_hh_cell_network/pop0[0]/v
single_hh_cell_network/pop0[1]/caConc
single_hh_cell_network/pop0[1]/caConcExt
single_hh_cell_network/pop0[1]/iCa
single_hh_cell_network/pop0[1]/iChannels
single_hh_cell_network/pop0[1]/iSyn
single_hh_cell_network/pop0[1]/spiking
single_hh_cell_network/pop0[1]/surfaceArea
single_hh_cell_network/pop0[1]/totSpecCap
single_hh_cell_network/pop0[1]/v
single_hh_cell_network/pop1/0/caConc
single_hh_cell_network/pop1/0/caConcExt
single_hh_cell_network/pop1/0/iCa
single_hh_cell_network/pop1/0/iChannels
single_hh_cell_network/pop1/0/iSyn
single_hh_cell_network/pop1/0/spiking
single_hh_cell_network/pop1/0/surfaceArea
single_hh_cell_network/pop1/0/totSpecCap
single_hh_cell_network/pop1/0/v
single_hh_cell_network/pop1/1/caConc
single_hh_cell_network/pop1/1/caConcExt
single_hh_cell_network/pop1/1/iCa
single_hh_cell_network/pop1/1/iChannels
single_hh_cell_network/pop1/1/iSyn
single_hh_cell_network/pop1/1/spiking
single_hh_cell_network/pop1/1/surfaceArea
single_hh_cell_network/pop1/1/totSpecCap
single_hh_cell_network/pop1/1/v
```

**Pull request merged**